**2. Hava durumu modeli**                                                         
Önceki derste olduğu gibi, hedef değişkeninizin görünür sıcaklık ve sıcaklık arasındaki fark olduğu bir doğrusal regresyon modeli oluşturun. Açıklayıcı değişkenler olarak nem ve rüzgar hızı kullanın. Şimdi, modelinizi OLS kullanarak tahmin edin. R-kare ve ayarlanmış R-kare değerleri nelerdir? Tatminkar olduklarını düşünüyor musunuz? Açıklayın.
Daha sonra, yukarıdaki modele nem ve rüzgar hızı etkileşimini dahil edin ve OLS'yi kullanarak modeli tahmin edin. Şimdi, bu modelin R-kare değeri nedir? Bu model bir öncekine göre gelişti mi?
İlk modele ek açıklayıcı değişken olarak görünürlük ekleyin ve tahmin edin. R-kare arttı mı? Ayarlanmış R-kare değeri ne oldu? Tabloda ortaya çıkan farklılıkları, ayarlanmış R-kare içindeki iyileşme açısından etkileşim terimi ve görünürlük ile karşılaştırın. Hangisi daha kullanışlı?
AIC ve BIC puanlarına göre yukarıdaki üç modelden en iyisini seçin. Mentor ile gerekçenizi tartışarak seçiminizi doğrulayın.

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
from sklearn import linear_model
import matplotlib.pyplot as plt
%matplotlib inline
import seaborn as sns
from sklearn import preprocessing
from scipy.stats.mstats import winsorize
import warnings
warnings.filterwarnings('ignore')

In [3]:
weather = pd.read_csv('C:\\Users\\Pc\\Desktop\\weatherHistory.csv')
weather.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

**Hedef Değişken**

In [4]:
weather['temp'] = weather['Temperature (C)'] - weather['Apparent Temperature (C)']

Doğrusal regresyon parametrelerini tahmin edelim.

**Model (Etkileşimsiz)**

In [5]:
X = weather[['Humidity','Wind Speed (km/h)']]
Y = weather[['temp']]
lrm = linear_model.LinearRegression()
lrm.fit(X, Y)
X = sm.add_constant(X)
results = sm.OLS(Y, X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   temp   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Tue, 09 Apr 2019   Prob (F-statistic):               0.00
Time:                        20:59:27   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Tahmin edilen modelde R kare değeri ve ayarlanmış R kare değeri 0.288'dir. Değer varyansın açıklanması için düşüktür.

**Model (Etkileşimli) Nem ve Rüzgar Hızı**

In [6]:
weather['hwind'] = weather['Humidity'] * weather['Wind Speed (km/h)']

X1 = weather[['Humidity', 'Wind Speed (km/h)', 'hwind']]
Y1 =  weather[['temp']]
lrm = linear_model.LinearRegression()
lrm.fit(X1, Y1)
X1 = sm.add_constant(X1)
result1 = sm.OLS(Y1, X1).fit()
result1.summary()


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   temp   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Tue, 09 Apr 2019   Prob (F-statistic):               0.00
Time:                        20:59:31   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity             -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)    -0.0905      0.002    -36.797      0.000      -0.095      -0.086
hwind                 0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Tahmin Modeli:

Sıcaklık=-0.0839-0.1775*nem-0.0905*rüzgarhızı+0.2971*hwind

Modelde tüm katsayılar istatistiksel olarak anlamlıdır. nem, rüzgar hızı ve etkileşimleri hissedilen sıcaklığı etkiler. R kare ve ayarlanmış R Kare 0.341 olmuştur. Bu modelimiz diğerine göre daha çok varyansı açıklıyor.

**Model(İlk Modele Görünürlük Dahil Edildi.)**

In [7]:
X2 = weather[['Humidity','Wind Speed (km/h)', 'Visibility (km)']]
Y2 = weather[['temp']]

X2 = sm.add_constant(X2)

result2 = sm.OLS(Y2,X2).fit()
result2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                   temp   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Tue, 09 Apr 2019   Prob (F-statistic):               0.00
Time:                        20:59:35   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -1.5756      0.028    -56.605      0.000      -1.630      -1.521
Humidity              2.6066      0.025    102.784      0.000       2.557       2.656
Wind Speed (km/h)     0.1199      0.001    179.014      0.000       0.119       0.121
Visibility (km)      -0.0540      0.001    -46.614      0.000      -0.056      -0.052
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                           0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R kare değeri 0.304 ilk modele göre yüksek ama ikinci modele göre düşük bir değerdir. İkinci model yani etkileşimli model hedef değişkenimizi daha iyi açıklıyor. İkinci model ve görünürlük modeli için de etkileşmli modelin R kare değeri daha fazladır. Etkileşimli model varyansı açıklamada bu modeller arasında en etkilisidir. 

İlk modelimiz için R-kare 0.288, ayarlanmış R-kare 0.288, AIC 3.409e+05 ve BIC 3.409e+05, ikinci modelimiz için ise R-kare 0.341, ayarlanmış R-kare 0.341, AIC 3.334e+05 ve BIC 3.334e+05, 3.modelimiz için R-kare 0.304, ayarlanmış R-kare 0.303, AIC 3.388e+05 ve BIC 3.388e+05'dir. Tüm metriklere göre, ikinci modelimiz diğerlerinden daha iyi görünüyor.

**3. Ev fiyatları modeli**                       
Ev fiyatları modelinizi tekrar çalıştırın ve F-testi, R-kare, ayarlanmış R-kare, AIC ve BIC kullanarak modelinizin uygunluğunu değerlendirin.
Modelinizin tatmin edici olduğunu düşünüyor musunuz? Açıklayın
Modelinizin uygunluğunu iyileştirmek için, bazı değişkenleri ekleyerek veya kaldırarak farklı model özelliklerini deneyin.
Denediğiniz her model için, uygun metrikleri alın ve modellerinizi birbiriyle karşılaştırın. Hangi model en iyisidir ve neden?

In [8]:
data = pd.read_csv('C:\\Users\\Pc\\Desktop\\train.csv')
data.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities    ...     PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
5         Lvl    AllPub    ...            0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
7         Lvl    AllPub    ...            0    NaN    NaN        Shed     350   
8         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
9         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

In [9]:
df=data.loc[:, ['Street', 'OverallQual', 'YearBuilt', 'ExterQual', 'SalePrice']]
df.head()

Street  OverallQual  YearBuilt ExterQual  SalePrice
0   Pave            7       2003        Gd     208500
1   Pave            6       1976        TA     181500
2   Pave            7       2001        Gd     223500
3   Pave            7       1915        TA     140000
4   Pave            8       2000        Gd     250000

In [10]:
df["Str"] = pd.get_dummies(df.Street, drop_first=True)
df["Ext"] = df.ExterQual.replace({'Gd':1, "TA":0, "Ex":2, "Fa":3})
df[['Str','Ext']].head()

Str  Ext
0    1    1
1    1    0
2    1    1
3    1    0
4    1    1

Modeldeki YearBuilt ve OverallQual ve SalePrice değişkenlerini standartlaştıralım.

In [28]:
from sklearn.preprocessing import scale
df['st_yearbuilt']=scale(df["YearBuilt"])
df['st_saleprice']=scale(df["SalePrice"])
df['st_ovqual']=scale(df["OverallQual"])
df.head()

Street  OverallQual  YearBuilt ExterQual  SalePrice  Str  Ext  st_yearbuilt  \
0   Pave            7       2003        Gd     208500    1    1      1.050994   
1   Pave            6       1976        TA     181500    1    0      0.156734   
2   Pave            7       2001        Gd     223500    1    1      0.984752   
3   Pave            7       1915        TA     140000    1    0     -1.863632   
4   Pave            8       2000        Gd     250000    1    1      0.951632   

   st_saleprice  st_ovqual  
0      0.347273   0.651479  
1      0.007288  -0.071836  
2      0.536154   0.651479  
3     -0.515281   0.651479  
4      0.869843   1.374795

**Model Tahmini**

In [30]:
Y3=df['st_saleprice']
X3=df[['Str', 'st_ovqual', 'st_yearbuilt', 'Ext']]
lrm = linear_model.LinearRegression()
lrm.fit(X3, Y3)
X3 = sm.add_constant(X3)
result1 = sm.OLS(Y3, X3).fit()
result1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           st_saleprice   R-squared:                       0.648
Model:                            OLS   Adj. R-squared:                  0.647
Method:                 Least Squares   F-statistic:                     668.3
Date:                Tue, 09 Apr 2019   Prob (F-statistic):               0.00
Time:                        23:54:07   Log-Likelihood:                -1310.4
No. Observations:                1460   AIC:                             2631.
Df Residuals:                    1455   BIC:                             2657.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -0.3884      0.250     -1.552      0.121      -0.879       0.102
Str              0.2801      0.248      1.130      0.258      -0.206       0.766
st_ovqual        0.6615      0.021     31.421      0.000       0.620       0.703
st_yearbuilt     0.0691      0.020      3.540      0.000       0.031       0.107
Ext              0.2520      0.033      7.747      0.000       0.188       0.316
==============================================================================
Omnibus:                      634.055   Durbin-Watson:                   1.971
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             7242.926
Skew:                           1.706   Prob(JB):                         0.00
Kurtosis:                      13.364   Cond. No.                         34.0
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Street katsayısı hariç diğer katsayılar istatistiksel olarak anlamlıdır. hedef değişkenini yani konut fiyatlarını bina yaşı, bina genel malzemesi ve dış kaplama malzemesi etkilemektedir. R kare değeri 0.648, ayarlanmış R kare 0.647'dir Aslında yüksek bir değer ama Street özelliğini çıkarıp bakalım.

Street değişkenini modelimizden çıkaralım ve yeni model için tahmin katsayılarını bulalım.

In [31]:
Y4 = df['st_saleprice']
X4 = df[['st_yearbuilt', 'Ext', 'st_ovqual' ]]
lrm = linear_model.LinearRegression()
lrm.fit(X4, Y4)
X4 = sm.add_constant(X4)
result1 = sm.OLS(Y4, X4).fit()
result1.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           st_saleprice   R-squared:                       0.647
Model:                            OLS   Adj. R-squared:                  0.647
Method:                 Least Squares   F-statistic:                     890.4
Date:                Tue, 09 Apr 2019   Prob (F-statistic):               0.00
Time:                        23:55:00   Log-Likelihood:                -1311.0
No. Observations:                1460   AIC:                             2630.
Df Residuals:                    1456   BIC:                             2651.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -0.1066      0.021     -5.107      0.000      -0.148      -0.066
st_yearbuilt     0.0697      0.020      3.572      0.000       0.031       0.108
Ext              0.2454      0.032      7.669      0.000       0.183       0.308
st_ovqual        0.6645      0.021     31.825      0.000       0.624       0.705
==============================================================================
Omnibus:                      630.404   Durbin-Watson:                   1.971
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             7207.573
Skew:                           1.693   Prob(JB):                         0.00
Kurtosis:                      13.345   Cond. No.                         3.18
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

Özellik değişkenleri model üzerinde pozitif yönde etkili. En çok etkili olan değişken bina yaşıdır. Anlamlı bir sonuç olduğunu söyleyebiliriz. R kare ve ayarlanmış R kare değeri bir önceki modele göre değişmedi ama AIC ve BIC değerleri 2630, 2651 bir önceki modele göre azalmıştır. 

Konut fiyatına etkili olduğunu düşündüğüm GrLivArea değişkenini modele dahil edelim

In [13]:
df1=data.loc[:, ['OverallQual', 'YearBuilt', 'ExterQual', 'SalePrice', 'GrLivArea']]
df1["Ext"] = df1.ExterQual.replace({'Gd':1, "TA":0, "Ex":2, "Fa":3})

In [33]:
df1['st_grliv']=scale(df1["GrLivArea"])
df1['st_yearbuilt']=scale(df1["YearBuilt"])
df1['st_saleprice']=scale(df1["SalePrice"])
df1['st_ovqual']=scale(df1["OverallQual"])
df1.head()

OverallQual  YearBuilt ExterQual  SalePrice  GrLivArea  Ext  st_grliv  \
0            7       2003        Gd     208500       1710    1  0.370333   
1            6       1976        TA     181500       1262    0 -0.482512   
2            7       2001        Gd     223500       1786    1  0.515013   
3            7       1915        TA     140000       1717    0  0.383659   
4            8       2000        Gd     250000       2198    1  1.299326   

   st_yearbuilt  st_saleprice  st_ovqual  
0      1.050994      0.347273   0.651479  
1      0.156734      0.007288  -0.071836  
2      0.984752      0.536154   0.651479  
3     -1.863632     -0.515281   0.651479  
4      0.951632      0.869843   1.374795

In [34]:
Y5 = df1['st_saleprice']
X5 = df1[['st_yearbuilt', 'Ext', 'st_ovqual', 'st_grliv' ]]
lrm = linear_model.LinearRegression()
lrm.fit(X5, Y5)
X4 = sm.add_constant(X5)
result2 = sm.OLS(Y5, X5).fit()
result2.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:           st_saleprice   R-squared:                       0.742
Model:                            OLS   Adj. R-squared:                  0.741
Method:                 Least Squares   F-statistic:                     1046.
Date:                Wed, 10 Apr 2019   Prob (F-statistic):               0.00
Time:                        00:00:18   Log-Likelihood:                -1083.3
No. Observations:                1460   AIC:                             2175.
Df Residuals:                    1456   BIC:                             2196.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
st_yearbuilt     0.1745      0.017     10.315      0.000       0.141       0.208
Ext              0.1017      0.020      4.966      0.000       0.062       0.142
st_ovqual        0.4153      0.021     19.995      0.000       0.375       0.456
st_grliv         0.4047      0.017     23.845      0.000       0.371       0.438
==============================================================================
Omnibus:                      387.774   Durbin-Watson:                   1.962
Prob(Omnibus):                  0.000   Jarque-Bera (JB):            15013.662
Skew:                           0.489   Prob(JB):                         0.00
Kurtosis:                      18.679   Cond. No.                         2.81
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

En son modelde p değerleri O'a yakın bütün katsayılar anlamlı R-kare 0.742, ayarlanmış R-kare 0.741 oldukça yüksek, AIC 2175 ve BIC 2196 düşük. Bu model diğerlerine göre oldukça iyi görünüyor. Konut fiyatlarını açıklamada en iyi model son model diyebiliriz.